kafka/main.go

```diff
@@ -67,82 +82,165 @@ func StartConsumers(providedServer *server.Server, logger *zerolog.Logger) error
 		topics = append(topics, topicMapping.Topic)
 	}
 
-	consumerCount, err := strconv.Atoi(os.Getenv("KAFKA_CONSUMERS_PER_NODE"))
-	if err != nil {
-		logger.Error().Err(err).Msg("Failed to convert KAFKA_CONSUMERS_PER_NODE variable to a usable integer. Defaulting to 1.")
-		consumerCount = 1
-	}
-
-	logger.Trace().Msgf("Spawning %d consumer goroutines", consumerCount)
+	reader := newConsumer(topics, adsConsumerGroupV1, logger)
 
-	for i := 1; i <= consumerCount; i++ {
-		go func(topicMappings []TopicMapping) {
-			consumer := newConsumer(topics, adsConsumerGroupV1, logger)
-			var (
-				failureCount = 0
-				failureLimit = 10
-			)
-			logger.Trace().Msg("Beginning message processing")
-			for {
-				// `FetchMessage` blocks until the next event. Do not block main.
-				ctx := context.Background()
-				logger.Trace().Msgf("Fetching messages from Kafka")
-				msg, err := consumer.FetchMessage(ctx)
-				if err != nil {
-					logger.Error().Err(err).Msg("")
-					if failureCount > failureLimit {
-						break
-					}
-					failureCount++
-					continue
+	// `kafka-go` exposes messages one at a time through its normal interfaces despite
+	// collecting messages with batching from Kafka. To process these messages in
+	// parallel we use the `FetchMessage` method in a loop to collect a set of messages
+	// for processing. Successes and permanent failures are committed. Temporary
+	// failures are not committed and are retried. Miscategorization of errors can
+	// cause the consumer to become stuck forever, so it's important that permanent
+	// failures are not categorized as temporary.
+	for {
+		var (
+			wg           sync.WaitGroup
+			errorResults = make(chan *utils.ProcessingError)
+		)
+		// Any error that occurs while getting the batch won't be available until
+		// the Close() call.
+		ctx := context.Background()
+		batch, err := batchFromReader(ctx, reader, 20, logger)
+		if err != nil {
+			logger.Error().Err(err).Msg("Batching failed")
+			// This should be an error that needs to communicate if its failure is
+			// temporary or permanent. If temporary we need to handle it and if
+			// permanent we need to commit and move on.
+		}
+	BatchProcessingLoop:
+		for _, msg := range batch {
+			wg.Add(1)
+			if err != nil {
+				// Indicates batch has no more messages. End the loop for
+				// this batch and fetch another.
+				if err == io.EOF {
```

Comment:
Is it possible that, if the batchFromReader errors with EOF and returns batch len(0) this will never get checked as the loop never starts so its kind of redundant or in case where the batchFromReader returns part of the required batch for example 2 messages and then also errors with EOF then those two messages will never be processed

Reply:
Good catch. This case is already handled in `batchFromReader` and this handler is vestigial. Removed.